cpp/include/raft/core/operators.hpp

```diff
@@ -0,0 +1,262 @@
+  template <typename T1, typename T2>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(T1 a, T2 b) const
+  {
+    return a * b;
+  }
+};
+
+struct div_op {
+  template <typename T1, typename T2>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(T1 a, T2 b) const
+  {
+    return a / b;
+  }
+};
+
+struct div_checkzero_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(Type a, Type b) const
+  {
+    if (b == Type{0}) { return Type{0}; }
+    return a / b;
+  }
+};
+
+struct pow_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(Type a, Type b) const
+  {
+    return std::pow(a, b);
+  }
+};
+
+struct min_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(Type a, Type b) const
+  {
+    if (a > b) { return b; }
+    return a;
+  }
+};
+
+struct max_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(Type a, Type b) const
+  {
+    if (b > a) { return b; }
+    return a;
+  }
+};
+
+struct sqdiff_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(Type a, Type b) const
+  {
+    Type diff = a - b;
+    return diff * diff;
+  }
+};
+
+struct argmin_op {
+  template <typename KVP>
+  constexpr RAFT_INLINE_FUNCTION KVP operator()(const KVP& a, const KVP& b) const
+  {
+    if ((b.value < a.value) || ((a.value == b.value) && (b.key < a.key))) { return b; }
+    return a;
+  }
+};
+
+struct argmax_op {
+  template <typename KVP>
+  constexpr RAFT_INLINE_FUNCTION KVP operator()(const KVP& a, const KVP& b) const
+  {
+    if ((b.value > a.value) || ((a.value == b.value) && (b.key < a.key))) { return b; }
+    return a;
+  }
+};
+
+template <typename ScalarT>
+struct const_op {
+  const ScalarT scalar;
+
+  constexpr const_op(ScalarT s) : scalar{s} {}
+
+  template <typename InT>
+  constexpr RAFT_INLINE_FUNCTION ScalarT operator()(InT unused) const
+  {
+    return scalar;
+  }
+};
+
+template <typename ComposedOpT, typename ScalarT>
+struct scalar_op {
+  ComposedOpT composed_op;
+  const ScalarT scalar;
+
+  constexpr scalar_op(ScalarT s) : scalar{s} {}
+
+  template <typename InT>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(InT a) const
+  {
+    return composed_op(a, scalar);
+  }
+};
+
+template <typename Type>
+using scalar_add_op = scalar_op<add_op, Type>;
+
+template <typename Type>
+using scalar_sub_op = scalar_op<sub_op, Type>;
+
+template <typename Type>
+using scalar_mul_op = scalar_op<mul_op, Type>;
+
+template <typename Type>
+using scalar_div_op = scalar_op<div_op, Type>;
+
+template <typename Type>
+using scalar_div_checkzero_op = scalar_op<div_checkzero_op, Type>;
+
+template <typename Type>
+using scalar_pow_op = scalar_op<pow_op, Type>;
+
+template <typename OuterOpT, typename InnerOpT>
+struct unary_compose_op {
+  OuterOpT outer_op;
+  InnerOpT inner_op;
+
+  template <typename... Args>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(Args... args) const
```

Comment:
Do you think we should care about having the perfect forwarding here? Certainly, most of the time we just pass arithmetic types here, but maybe once in a while a struct passed by reference could come up. Hence passing it by value wouldn't be optimal.

Reply:
What I have observed so far is that if the inner op takes references, as `Args...` are inferred as references, structs are passed through with zero copy. But it doesn't hurt to add `std::forward` so I will. If you have an example of a case where that makes a difference, I'd be curious to read it.

Consider the following example, the output is 37:

```c++
template <typename T>
void f(T& a, const T& b) {
    a = b;
}

template<typename... Args>
void g(Args... args) {
    f(args...);
}

int main() {
    int a = 13;
    f(a, 37);
    std::cout << a << std::endl;
}
```

Reply:
You're right, we need perfect forwarding here to pass args by reference, I'll add that.